source/dagger/gameplay/ping_pong/pingpong_playerinput.cpp

```diff
@@ -23,34 +28,107 @@ void PingPongPlayerInputSystem::WindDown()
+            }
+            else if (kEvent_.key == ctrl_.up_key && kEvent_.action == EDaggerInputState::Released && ctrl_.input.y > 0)
+            {
+                ctrl_.input.y = 0;
+            }
+            else if (kEvent_.key == ctrl_.down_key && (kEvent_.action == EDaggerInputState::Held || kEvent_.action == EDaggerInputState::Pressed))
+            {
+                ctrl_.input.y = -1;
+            }
+            else if (kEvent_.key == ctrl_.down_key && kEvent_.action == EDaggerInputState::Released && ctrl_.input.y < 0)
+            {
+                ctrl_.input.y = 0;
+            }
+            else if (kEvent_.key == ctrl_.slow_down_key && kEvent_.action == EDaggerInputState::Pressed) {
+                if (pwrups_.slow_down > 0 && s_PowerUpActive == false && s_ActivatePowerUp == false) {
+                    pwrups_.power_up = true;
+                    s_ActivatePowerUp = true;
+                }
+
+                else if (s_PowerUpActive == true && pwrups_.power_up == true){
+                    pwrups_.power_up = false;
+                    pwrups_.power_down = true;
+                    s_DeactivatePowerUp = true;
+                }
+            }
+    };
 }
 
 void PingPongPlayerInputSystem::Run()
 {
-    auto view = Engine::Registry().view<Transform, ControllerMapping>();
+    auto view = Engine::Registry().view<Transform, ControllerMapping, PlayerPowerUp>();
     for (auto entity : view)
     {
-        auto &t = view.get<Transform>(entity);
-        auto &ctrl = view.get<ControllerMapping>(entity);
+        auto& t = view.get<Transform>(entity);
+        auto& ctrl = view.get<ControllerMapping>(entity);
+        auto& pwrup = view.get<PlayerPowerUp>(entity);
+
+        //If power-up has been active for more than 5s , then it deactivates automaticaly
+        if (s_PowerUpActive == true) {
+            s_TimeUnitlNextChange -= dagger::Engine::Instance().DeltaTime();
+            if(s_TimeUnitlNextChange <= 0){
+                Engine::Registry().view<PlayerPowerUp>().each([&](PlayerPowerUp& pwrup) {
+                    if (pwrup.power_up == true) {
+                        pwrup.power_up = false;
+                        pwrup.power_down = true;
+                        s_DeactivatePowerUp = true;
+                    }
+                
+                });
+                s_TimeUnitlNextChange = s_PowerUpDuration;
+            }
+            
+        }
+
+        if (pwrup.power_up == true && s_PowerUpActive == false && s_ActivatePowerUp == true) {
+                pwrup.slow_down--;
+            
+                s_PlayerSpeed = s_PlayerSpeed * 0.5f;
+                s_PowerUpActive = true;
+                s_ActivatePowerUp = false;
+
+                Engine::Registry().view<PingPongBall>().each([&](PingPongBall& ball) {
+
+                    if (ball.reachedGoal)
+                    {
+                        return;
+                    }
+
+                    ball.speed.x = ball.speed.x * 0.33f;
```

Comment:
it'll be good to store 0.33f as a const somewhere, so it'll be easier to change it later.

Reply:
Added two consts , for ball and player speed multipliers, as you suggested